docs/source/grammar.rst

```diff
@@ -0,0 +1,240 @@
+.. testsetup:: *
+
+    from omegaconf import OmegaConf
+
+The OmegaConf grammar
+---------------------
+
+OmegaConf uses an `ANTLR <https://www.antlr.org/>`_-based grammar to parse string expressions,
+where the `lexer rules <https://github.com/omry/omegaconf/blob/master/omegaconf/grammar/OmegaConfGrammarLexer.g4>`_
+rules define the tokens used by the `parser rules <https://github.com/omry/omegaconf/blob/master/omegaconf/grammar/OmegaConfGrammarParser.g4>`_.
+Currently this grammar's main usage is in the parsing of :ref:`interpolations<interpolation>`, detailed below.
+
+
+String interpolations
+^^^^^^^^^^^^^^^^^^^^^
+
+A string interpolation is a string concatenating an interpolation with additional characters
+(that may or may not be other interpolations).
+These are all examples of string interpolations:
+
+    - ``https://${host}:${port}``
+    - ``Hello ${name}``
+    - ``${a}${oc.env:B}${c}``
+
+Such strings are matched by the following parser rule, which is the main rule used
+to parse any string containing an interpolation:
+
+    .. code-block:: antlr
+
+        text: (interpolation |
+               ANY_STR | ESC | ESC_INTER | TOP_ESC | QUOTED_ESC)+;
+
+In this rule, all tokens except ``interpolation`` represent regular string fragments
+(with a special handling of escaped characters, see :ref:`escaping-in-sring-interpolations` below).
+
+
+Interpolation types
+^^^^^^^^^^^^^^^^^^^
+
+An ``interpolation`` as found in the rule above can either be a :ref:`config-node-interpolation`
+(e.g., ``${host}``) or a call to a :ref:`resolver<resolvers>` (e.g., ``${oc.env:B}``).
+This is reflected in the following parser rules:
+
+    .. code-block:: antlr
+
+        interpolation: interpolationNode | interpolationResolver;
+
+        interpolationNode:
+            INTER_OPEN  // ${
+            DOT* 
+            (configKey | BRACKET_OPEN configKey BRACKET_CLOSE)
+            (DOT configKey | BRACKET_OPEN configKey BRACKET_CLOSE)*
+            INTER_CLOSE;  // }
+
+        interpolationResolver:
+            INTER_OPEN  // ${
+            resolverName COLON sequence?
+            BRACE_CLOSE;  // }
+
+The following are all valid examples of config node interpolations according to the ``interpolationNode`` rule
+(note in particular that it supports both dot and bracket notations to access child nodes):
+
+    - ``${host}``
+    - ``${.sibling}``
+    - ``${..uncle.cousin}``
+    - ``${some_list[3]}``
+    - ``${some_deep_dict[key1][subkey2].subsubkey3}``
```

Comment:
I can see how this may look a bit confusing. The underlying reason is that the leading dots notation for relative interpolations is "special" (those dots aren't the same as the dots used to separate a parent node from its child).

For instance `${.}` denotes the parent node (this isn't allowed anymore because it can lead to cycles in the config, but it used to work at some point). So if you think as a leading dot as meaning "parent", `${.foo}` means `${parent.foo}` and `${.[foo]}` would mean `${parent[foo]}`.

I'll let @omry chime in if he thinks we should change the current behavior. Personally I would keep it, because I like that relative interpolations are consistently using leading dots (if `${[key]}` became a relative interpolation we would lose this consistency).

Reply:
This explanation makes sense to me.
Also, I think that if we make `${[key]}` invalid then there will be no easy way to access the root config with brackets notation.

Reply:
This explanation makes sense to me. The "parental" dots at the beginning of the interpolation have a different meaning from the "getattr" dots that appear between keys.